tough/src/transport.rs

```diff
@@ -25,61 +26,97 @@ dyn_clone::clone_trait_object!(Transport);
     FileNotFound,
-    /// The transport failed for any other reason, e.g. IO error, HTTP broken pipe, etc.
+    /// The transport failed for any other reason, e.g. IO error, HTTP broken pipe, etc. The
+    /// implication is that the [`Transport`] the file might exist, but the [`Transport`] failed.
     Other,
 }
 
-/// The error type that [`Transport`] `fetch` returns.
-#[derive(Debug, Snafu)]
-#[snafu(display("{:?} error fetching '{}': {}", kind, url, source))]
+/// The error type that [`Transport::fetch`] returns.
+#[derive(Debug)]
 pub struct TransportError {
     /// The kind of error that occurred.
-    pub kind: TransportErrorKind,
+    kind: TransportErrorKind,
     /// The URL that the transport was trying to fetch.
-    pub url: String,
-    /// The underlying error that occurred.
-    pub source: Box<dyn std::error::Error + Send + Sync>,
+    url: String,
+    /// The underlying error that occurred (if any).
+    source: Option<Box<dyn Error + Send + Sync>>,
 }
 
 impl TransportError {
-    /// Creates a new [`TransportError`].
-    pub fn new<S, E>(kind: TransportErrorKind, url: S, source_error: E) -> Self
+    /// Creates a new [`TransportError`]. Use this when there is no underlying error to wrap.
+    pub fn new<S>(kind: TransportErrorKind, url: S) -> Self
     where
-        E: Into<Box<dyn std::error::Error + Send + Sync>>,
         S: AsRef<str>,
     {
         Self {
             kind,
             url: url.as_ref().into(),
-            source: source_error.into(),
+            source: None,
         }
     }
 
-    /// Creates a [`TransportError`] for reporting an unhandled URL type.
-    pub fn unsupported_scheme<S: AsRef<str>>(url: S) -> Self {
-        TransportError::new(
-            TransportErrorKind::UnsupportedUrlScheme,
-            url,
-            "Transport cannot handle the given URL scheme.".to_string(),
-        )
+    /// Creates a new [`TransportError`]. Use this to preserve an underlying error.
+    pub fn new_with_cause<S, E>(kind: TransportErrorKind, url: S, source: E) -> Self
+    where
+        E: Into<Box<dyn Error + Send + Sync>>,
+        S: AsRef<str>,
+    {
+        Self {
+            kind,
+            url: url.as_ref().into(),
+            source: Some(source.into()),
+        }
+    }
+
+    /// The type of [`Transport`] error that occurred.
+    pub fn kind(&self) -> TransportErrorKind {
+        self.kind
+    }
+
+    /// The URL that the [`Transport`] was trying to fetch when the error occurred.
+    pub fn url(&self) -> &str {
+        self.url.as_str()
+    }
+}
+
+impl Display for TransportError {
+    fn fmt(&self, f: &mut Formatter<'_>) -> std::fmt::Result {
+        if let Some(e) = self.source.as_ref() {
+            write!(
+                f,
+                "Transport {:?} error fetching '{}': {}",
+                self.kind, self.url, e
+            )
+        } else {
+            write!(f, "Transport {:?} error fetching '{}'", self.kind, self.url)
```

Comment:
IMO would be better to `impl Display for TransportErrorKind` and use that friendlier version here.